### PR TITLE
Add debug_traceTransaction

### PR DESCRIPTION
**Pull request by dongrie**
When an error occurs in a transaction, we use txHash to output `revertReason` `contract` in the error message with debug trace.
This feature can be controlled on or off in ChainConfig.

# Before
before

# After
after

# Not debug mode
geth-error

### CODE REVIEW

File: pkg/client/client.go [hunk 1]
```diff
@@ -70,15 +70,24 @@ func (cl *ETHClient) GetTransactionReceipt(ctx context.Context, txHash common.Ha
 		reason, err := r.GetRevertReason()
 		return &r.Receipt, false, fmt.Errorf("revert: %v(parse-err=%v)", reason, err)
 	} else {
-		return &r.Receipt, false, fmt.Errorf("failed to execute a transaction: %v", r)
+		errPrefix := "failed to execute a transaction"
+		if enableDebugTrace {
+			to, revertReason, err := cl.DebugTraceTransaction(ctx, txHash)
+			if err != nil {
+				return &r.Receipt, false, fmt.Errorf("%s: %v, debug_transaction error: %v", errPrefix, r, err)
```
Comment (toshihiko-okubo) [inline pkg/client/client.go:77]: Is it correct to return ERROR when Tx is successful?

**Review by toshihiko-okubo — APPROVED** [review]

File: pkg/client/client.go [hunk 2]
```diff
@@ -70,15 +70,24 @@ func (cl *ETHClient) GetTransactionReceipt(ctx context.Context, txHash common.Ha
...
+			}
+			return &r.Receipt, false, fmt.Errorf("%s: %v, contract: %s, revert reason: %v", errPrefix, r, to, revertReason)
```
Comment (toshihiko-okubo) [inline pkg/client/client.go:79]: could you add TxHash to the log

File: pkg/client/client.go [hunk 3]
```diff
@@ -126,3 +151,30 @@ func parseRevertReason(bz []byte) (string, error) {
...
+}
+
+func searchToAndReason(to string, calls []Result) (string, string) {
+	for _, call := range calls {
+		if call.To != nil {
+			to = *call.To
+		}
+		if call.RevertReason != nil {
+			return to, *call.RevertReason
+		}
```
Comment (siburu) [inline pkg/client/client.go:176]: Can there be a case where `call.RevertReason != nil` but `call.To == nil`?
If any, when does it happen?

Comment (siburu) [reply pkg/client/client.go:176]: I think the following function will make the code more natural.

```
func searchToAndReason(result Result) (string, string, error) {
    if result.RevertReason != nil {
        return *result.To, *result.RevertReason, nil
    }
    for _, call := range result.Calls {
        to, reason, err := searchToAndReason(call)
        if err == nil {
            return to, reason, nil
        }
    }
    return "", "", fmt.Errorf("revert reason not found")
}
```

Comment (dongrie) [reply pkg/client/client.go]: Fixed search reason logic
https://github.com/datachainlab/ethereum-ibc-relay-chain/pull/10/commits/53a99c31942d630704e8abb75c5d2899fe179aec

File: pkg/client/client.go (hunk 2 above)
Comment (dongrie) [reply pkg/client/client.go]: Screenshot 2023-08-01 at 13 23 43
Add txHash to the log
https://github.com/datachainlab/ethereum-ibc-relay-chain/pull/10/commits/53a99c31942d630704e8abb75c5d2899fe179aec

**Pull request merged**